File: app/src/main/java/com/afoxxvi/alopex/ui/dialog/FilterDialog.java
package com.afoxxvi.alopex.ui.dialog;

import android.annotation.SuppressLint;
import android.content.Context;
import android.view.LayoutInflater;
import android.widget.EditText;
import android.widget.Toast;

import com.afoxxvi.alopex.AlopexView;
import com.afoxxvi.alopex.component.filter.AlopexFilter;
import com.afoxxvi.alopex.component.filter.AlopexFilterManager;
import com.afoxxvi.alopex.databinding.DialogFilterBinding;
import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class FilterDialog extends BaseDialog {
    private final DialogFilterBinding binding;
    private final AlopexFilter filter;
    private final List<AlopexFilter.Match> matchList;

    public FilterDialog(Context context, AlopexFilter filter) {
        super(context, "Edit filter");
        this.filter = filter;
        this.matchList = new ArrayList<>();
        for (AlopexFilter.Match match : filter.getMatchList()) {
            this.matchList.add(AlopexFilter.Match.copyOf(match));
        }
        binding = DialogFilterBinding.inflate(LayoutInflater.from(context));
        setContent(binding.getRoot());

        binding.editPackage.setText(filter.getPackageName());
        binding.editPackage.setOnClickListener(v -> new AppListDialog(context) {
            @Override
            public void onSelect(AppInfo appInfo) {
                super.onSelect(appInfo);
                binding.editPackage.setText(appInfo.getPackageName());
                dialog.dismiss();
            }
        }.show());
        binding.checkCancel.setChecked(filter.getCancelFiltered());
        binding.checkNotify.setChecked(filter.getNotifyUnfiltered());
        for (AlopexFilter.Match match : matchList) {
            addChip(binding.chipsBlacklist, match, matchList);
        }
        binding.buttonAddBlacklist.setOnClickListener(v -> addMatchRule(context, binding.editNewBlacklist, matchList, binding.chipsBlacklist, "enter match name."));
    }

    private void addMatchRule(Context context, EditText editFrom, List<AlopexFilter.Match> matchList, ChipGroup targetGroup, String emptyMsg) {
        String text = editFrom.getText().toString();
        if (text.length() > 0) {
            AlopexFilter.Match match = new AlopexFilter.Match(text);
            editFrom.setText("");
            new MatchRuleDialog(context, match) {
                @Override
                public void onConfirm() {
                    super.onConfirm();
                    matchList.add(match);
                    FilterDialog.this.addChip(targetGroup, match, matchList);
                }
            }.show();
        } else {
            AlopexView.showToast(context, emptyMsg, Toast.LENGTH_SHORT);
            editFrom.requestFocus();
        }
    }

    @Override
    public void onCancel() {
        super.onCancel();
    }

    @Override
    public void onConfirm() {
        AlopexFilterManager manager = AlopexFilterManager.getInstance();
        String pkgName = binding.editPackage.getText().toString();
        if (pkgName.isEmpty()) {
            binding.editPackage.requestFocus();
            AlopexView.showToast(context, "package name is empty.", Toast.LENGTH_SHORT);
            return;
        }
        filter.setPackageName(pkgName);
        filter.setMatchList(matchList);
        filter.setCancelFiltered(binding.checkCancel.isChecked());
        filter.setNotifyUnfiltered(binding.checkNotify.isChecked());
        filter.notifyChange();
        if (!manager.getFilters().contains(filter)) {
            manager.getFilters().add(filter);
        }
        manager.save(context);
        this.dialog.dismiss();
    }

    @SuppressLint("DefaultLocale")
    private void addChip(ChipGroup target, AlopexFilter.Match match, List<AlopexFilter.Match> list) {
        Chip chip = new Chip(context);
        chip.setText(String.format("%s(%d)", match.getName(), match.getMatchCount()));
        chip.setOnClickListener(v -> chip.setCloseIconVisible(!chip.isCloseIconVisible()));
        chip.setOnLongClickListener(v -> {
            new MatchRuleDialog(context, match).show();
            return true;
        });
        chip.setOnCloseIconClickListener(v ->
                new BaseDialog(context, "Delete match rule?") {
                    @Override
                    public void onConfirm() {
                        for (Iterator<AlopexFilter.Match> iterator = list.listIterator(); iterator.hasNext(); ) {
                            AlopexFilter.Match m = iterator.next();
                            if (m == match) {
                                iterator.remove();
                                target.removeView(chip);
                                break;
                            }
                        }
                        super.onConfirm();
                    }
                }.setButtonText("Yes", "No").show());
        target.addView(chip);
    }

    @Override
    public void show() {
        super.show();
        dialog.setCancelable(false);
    }
}

File: app/src/main/java/com/afoxxvi/alopex/ui/dialog/FilterDialog.kt
package com.afoxxvi.alopex.ui.dialog

import android.annotation.SuppressLint
import android.content.Context
import android.view.LayoutInflater
import android.widget.EditText
import android.widget.Toast
import com.afoxxvi.alopex.Alopex
import com.afoxxvi.alopex.component.filter.AlopexFilter
import com.afoxxvi.alopex.component.filter.AlopexFilterManager
import com.afoxxvi.alopex.databinding.DialogFilterBinding
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup

open class FilterDialog(context: Context, private val filter: AlopexFilter) : BaseDialog(context, "Edit filter") {
    private val binding: DialogFilterBinding
    private val matchList: MutableList<AlopexFilter.Match>

    init {
        matchList = ArrayList()
        filter.matchList!!.forEach { matchList.add(AlopexFilter.Match.copyOf(it)) }
        binding = DialogFilterBinding.inflate(LayoutInflater.from(context))
        setContent(binding.root)
        binding.editPackage.setText(filter.packageName)
        binding.editPackage.setOnClickListener {
            object : AppListDialog(context) {
                override fun onSelect(appInfo: AppInfo) {
                    super.onSelect(appInfo)
                    binding.editPackage.setText(appInfo.packageName)
                    dialog?.dismiss()
                }
            }.show()
        }
        binding.checkCancel.isChecked = filter.cancelFiltered
        binding.checkNotify.isChecked = filter.notifyUnfiltered
        matchList.forEach { addChip(binding.chipsBlacklist, it, matchList) }
        binding.buttonAddBlacklist.setOnClickListener { addMatchRule(context, binding.editNewBlacklist, matchList, binding.chipsBlacklist, "enter match name.") }
    }

    private fun addMatchRule(context: Context, editFrom: EditText, matchList: MutableList<AlopexFilter.Match>, targetGroup: ChipGroup, emptyMsg: String) {
        val text = editFrom.text.toString()
        if (text.isNotEmpty()) {
            val match = AlopexFilter.Match(text)
            editFrom.setText("")
            object : MatchRuleDialog(context, match) {
                override fun onConfirm() {
                    super.onConfirm()
                    matchList.add(match)
                    this@FilterDialog.addChip(targetGroup, match, matchList)
                }
            }.show()
        } else {
            Alopex.showToast(context, emptyMsg, Toast.LENGTH_SHORT)
            editFrom.requestFocus()
        }
    }

    override fun onConfirm() {
        val manager = AlopexFilterManager
        val pkgName = binding.editPackage.text.toString()
        if (pkgName.isEmpty()) {
            binding.editPackage.requestFocus()
            Alopex.showToast(context, "package name is empty.", Toast.LENGTH_SHORT)
            return
        }
        filter.packageName = pkgName
        filter.matchList = matchList
        filter.cancelFiltered = binding.checkCancel.isChecked
        filter.notifyUnfiltered = binding.checkNotify.isChecked
        filter.notifyChange()
        if (!manager.filters.contains(filter)) {
            manager.filters.add(filter)
        }
        manager.save(context)
        dialog?.dismiss()
    }

    @SuppressLint("DefaultLocale")
    private fun addChip(target: ChipGroup, match: AlopexFilter.Match, list: MutableList<AlopexFilter.Match>) {
        val chip = Chip(context)
        chip.text = String.format("%s(%d)", match.name, match.matchCount)
        chip.setOnClickListener { chip.isCloseIconVisible = !chip.isCloseIconVisible }
        chip.setOnLongClickListener {
            MatchRuleDialog(context, match).show()
            true
        }
        chip.setOnCloseIconClickListener {
            object : BaseDialog(context, "Delete match rule?") {
                override fun onConfirm() {
                    val iterator: MutableIterator<AlopexFilter.Match> = list.listIterator()
                    while (iterator.hasNext()) {
                        val m = iterator.next()
                        if (m === match) {
                            iterator.remove()
                            target.removeView(chip)
                            break
                        }
                    }
                    super.onConfirm()
                }
            }.setButtonText("Yes", "No").show()
        }
        target.addView(chip)
    }

    override fun show() {
        super.show()
        dialog?.setCancelable(false)
    }
}